src/type/validate.js

```diff
@@ -185,8 +204,129 @@ function validateDirectives(context: SchemaValidationContext): void {
+
+      if (arg.defaultValue) {
+        validateDefaultValue(
+          context,
+          `@${directive.name}(${arg.name}:)`,
+          arg.type,
+          arg.defaultValue,
+          arg.astNode?.defaultValue,
+        );
+      }
+    }
+  }
+}
+
+function validateDefaultValue(
+  context: SchemaValidationContext,
+  coordinate: string,
+  type: GraphQLInputType,
+  defaultValue: GraphQLDefaultValueUsage,
+  defaultValueNode: ?ConstValueNode,
+): void {
+  if (defaultValue.literal) {
+    validateInputLiteral(
+      defaultValue.literal,
+      type,
+      undefined,
+      (error, path) => {
+        context.reportError(
+          `${coordinate} has invalid default value${printPathArray(path)}: ${
+            error.message
+          }`,
+          error.nodes,
+        );
+      },
+    );
+  } else {
+    let errors = [];
+    validateInputValue(defaultValue.value, type, (error, path) => {
+      errors.push([error, path]);
+    });
+
+    // If there were validation errors, check to see if it can be "uncoerced"
+    // and then correctly validated. If so, report a clear error with a path
+    // to resolution.
+    if (errors.length > 0) {
+      try {
+        const uncoercedValue = uncoerceDefaultValue(defaultValue.value, type);
+        const uncoercedErrors = [];
+        validateInputValue(uncoercedValue, type, (error, path) => {
+          uncoercedErrors.push([error, path]);
+        });
+        if (uncoercedErrors.length === 0) {
+          context.reportError(
+            `${coordinate} has invalid default value: ${inspect(
+              defaultValue.value,
+            )}. Did you mean: ${inspect(uncoercedValue)}?`,
+            defaultValueNode,
+          );
+          errors = [];
+        }
+      } catch (_error) {
+        // ignore
+      }
+    }
+
+    for (const [error, path] of errors) {
+      context.reportError(
+        `${coordinate} has invalid default value${printPathArray(path)}: ${
+          error.message
+        }`,
+        defaultValueNode,
+      );
+    }
+  }
+}
+
+/**
+ * Historically GraphQL.js allowed default values to be provided as
+ * assumed-coerced "internal" values, however default values should be provided
+ * as "external" pre-coerced values. `uncoerceDefaultValue()` will convert such
+ * "internal" values to "external" values to display as part of validation.
+ *
+ * This performs the "opposite" of `coerceInputValue()`. Given an "internal"
+ * coerced value, reverse the process to provide an "external" uncoerced value.
+ */
+function uncoerceDefaultValue(value: mixed, type: GraphQLInputType): mixed {
+  if (isNonNullType(type)) {
+    return uncoerceDefaultValue(value, type.ofType);
+  }
+
+  if (value === null) {
+    return null;
+  }
+
+  if (isListType(type)) {
+    if (isIterableObject(value)) {
+      return Array.from(value, (itemValue) =>
+        uncoerceDefaultValue(itemValue, type.ofType),
+      );
     }
+    return uncoerceDefaultValue(value, type.ofType);
```

Comment:
I think this is doing the implicit array coercion; but since we're suggesting an improvement to the user should we include this array conversion explicitly?

```suggestion
    return [uncoerceDefaultValue(value, type.ofType)];
```